### ISSUE TITLE
Non-integer Scale degree calculations are incorrect

### ISSUE THREAD

**Issue by scztt** [supercollider/supercollider#4615]
## Steps to reproduce

```supercollider
// Common use-case: non-integral degrees in a pattern
Pbind(
	\degree, Pseq(Array.series(20, 0, 0.1)),
).asStream.nextN(20, Event.parentEvents.default)
.collect({
	|e|
	e.use { e.note }
});
// -> [ 0.0, 1.0, 2.0, 3.0, 4.0, -3.0, -2.0, -1.0, 4.4408920985006e-16, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, -1.0, 8.8817841970013e-16, 1.0, 2.0, 3.0 ]


// Reduced case:
[0, 0.5, 1].degreeToKey(#[0, 2, 4, 5, 7, 9, 11])
// -> [ 0, -3.0, 2 ]
```

## Expected vs. actual behavior
Fractional degrees should fall somewhere between the note values of the previous and next degree values (e.g. `3.0.degreeToKey(scale) < 3.5.degreeToKey < 4.0.degreeToKey`).

The code at fault is here:
```	degreeToKey { |scale, stepsPerOctave = 12|
		var scaleDegree = this.round.asInteger;
>>>		var accidental = (this - scaleDegree) * 10.0;
		^scale.performDegreeToKey(scaleDegree, stepsPerOctave, accidental)
	}
```
Elsewhere, accidental is treated as a simple multiplier rather than one normalized to (0..10):
```
		^if(accidental == 0) { baseKey } { baseKey + (accidental * (stepsPerOctave / 12.0)) }
```
There's an additional problem in `performDegreeToKey` implementations, which seem to account for the `accidental` incorrectly. I *believe* something like this (or something close to it....) should hold true, but even if the 10x multiplier is removed, this still does not hold:
```
3.5.degreeToKey(scale) == 3.degreeToKey(scale).blend(0.5, 4.degreeToKey(scale))
```

**Comment by jamshark70** [comment]
> Fractional degrees should fall somewhere between the note values of the previous and next degree values.

Actually, assuming a C scale, 0.1 is considered to be C#, 0.2 is C##, 0.3 is C### (and 0.5 is undefined, because it may be equally C##### or Dbbbbb... but the musical use case for quintuple sharps or flats is... well... rather staggering so we haven't bothered to resolve that).

So it is actually *not* expected that `3.0.degreeToKey(scale) < 3.5.degreeToKey < 4.0.degreeToKey`. 0.3, as a triple sharp, *should* resolve higher than degree 1.

There may be other bugs (in particular, I'm not thrilled that accidentals assume 12 semitones per octave) but I believe this is not one.

**Comment by jamshark70** [comment]
> 3.5.degreeToKey(scale) == 3.degreeToKey(scale).blend(0.5, 4.degreeToKey(scale))

And (overlooked this at first) -- I did this in my ModalSpec class, but it turned out to be a terrible idea. I recently added an option to ModalSpec methods to use SC's concept of accidentals!

Suppose you have a harmonic minor scale (0, 2, 3, 5, 7, 8, 11).

If you want to sharp the tonic, by your proposal you would add 0.5. To sharp the second, you would add 1. To sharp the 6th, you would add 1/3 (!).

In other words, in that case, the concept of raising or lowering a note by a semitone is insufficiently abstracted away from the form of the scale.

I found it eventually to be unworkable for musical logic and had to abandon it. James McCartney was right to implement accidentals as he did (except the 12 per octave assumption, but I don't have a solution for that).

**Comment by sonoro1234** [comment]
In Lua2SC I am using linear proportion to find fractional degrees
(1.5 as a degree would be halfway between 1 and 2)

with harmonic minor scale (0, 2, 3, 5, 7, 8, 11)
5+1/3==5.333333 degree would be calculated with a linear proportion to be note 9
5+2/3==5.666666 degree would be calculated with a linear proportion to be note 10
5+3/3==6 degree would be calculated with a linear proportion to be note 11

**Comment by jamshark70** [comment]
The problem I have with the linear proportion approach is: if I start with some arbitrary scale degree and I want to raise or lower it by exactly one semitone, I have to query the scale to find out the size of the interval. So the musical operation ("sharp/flat") is tightly coupled to the scale form.

Not only that operation. If I have a sharp 6th in harmonic minor (5.333333) and I want the note a perfect 4th below that, you would think the math operation would be `-3` but that gives you 2.333333 = minor 3rd above tonic + 2/3 of a semitone. So somehow, we have to find out that the correct amount to subtract is 2.8333333 (but if it's a normal 6th degree, P4 down is -3 as expected).

With proportions, it's easy to *write* pitches with accidentals. It's hard to *manipulate* pitches with accidentals.

This is one where I thought I could improve on JMc's design (by proportionally splitting intervals) but in the end, I could never make accidentals work in my live coding system that way. So I hadn't improved anything.

**Comment by brianlheim** [comment]
Closing as intended behavior. The truth here is probably somewhere in between "this is wrong" and "this can't be improved upon" -- the existence of wholly unintuitive results like this is evidence enough of that, but at least there is some consistency.

**Comment by jamshark70** [comment]
Last comment, then I'll leave it.

It *could* be improved upon by a ScaleDegree class, which has separate values for degree (integer) and accidentals (also integer), and implements math operators. Syntax is a problem (writing literal sequences would become extremely verbose).

The current approach is definitely a kludge. I'm especially displeased with the way accidentals are 12-ET centric, and (because you can go only +/-4) you can't use accidentals to address specific ET divisions in, say, 53ET. To make the kludge more useful, it should have been x.01 for one sharp rather than x.1.

But it would break code to change that.

If there is an expectation of proportional division of scale degrees, then we need to update the documentation to explain why that's not actually ideal.

Ok, I'll stop now.
